src/Type/TypeUtils.php

```diff
@@ -170,6 +170,24 @@ public static function findThisType(Type $type): ?ThisType
 		return null;
 	}
 
+	public static function findCallableType(Type $type): ?Type
+	{
+		if ($type->isCallable()->yes()) {
+			return $type;
+		}
+
+		if ($type instanceof UnionType || $type instanceof IntersectionType) {
```

Comment:
I don't think we need to go into intersections. Once one of the intersected types is callable, the whole type is callable.